invokeai/app/services/config.py

```diff
@@ -0,0 +1,364 @@
+        env_file_encoding = 'utf-8'
+        arbitrary_types_allowed = True
+        env_prefix = 'INVOKEAI_'
+        case_sensitive = True
+        @classmethod
+        def customise_sources(
+            cls,
+            init_settings,
+            env_settings,
+            file_secret_settings,
+        ):
+            return (
+                init_settings,
+                cls._omegaconf_settings_source,
+                env_settings,
+                file_secret_settings,
+            )
+
+        @classmethod
+        def _omegaconf_settings_source(cls, settings: BaseSettings) -> dict[str, Any]:
+            if initconf := InvokeAISettings.initconf:
+                return initconf.get(settings.cmd_name(),{})
+            else:
+                return {}
+
+    @staticmethod
+    def add_field_argument(command_parser, name: str, field, default_override = None):
+        default = default_override if default_override is not None else field.default if field.default_factory is None else field.default_factory()
+        if get_origin(field.type_) == Literal:
+            allowed_values = get_args(field.type_)
+            allowed_types = set()
+            for val in allowed_values:
+                allowed_types.add(type(val))
+            allowed_types_list = list(allowed_types)
+            field_type = allowed_types_list[0] if len(allowed_types) == 1 else Union[allowed_types_list]  # type: ignore
+
+            command_parser.add_argument(
+                f"--{name}",
+                dest=name,
+                type=field_type,
+                default=default,
+                choices=allowed_values,
+                help=field.field_info.description,
+            )
+        else:
+            command_parser.add_argument(
+                f"--{name}",
+                dest=name,
+                type=field.type_,
+                default=default,
+                action=argparse.BooleanOptionalAction if field.type_==bool else 'store',
+                help=field.field_info.description,
+            )
+def _find_root()->Path:
+    if os.environ.get("INVOKEAI_ROOT"):
+        root = Path(os.environ.get("INVOKEAI_ROOT")).resolve()
+    elif (
+            os.environ.get("VIRTUAL_ENV")
+            and (Path(os.environ.get("VIRTUAL_ENV"), "..", INIT_FILE).exists()
+                 or
+                 Path(os.environ.get("VIRTUAL_ENV"), "..", LEGACY_INIT_FILE).exists()
+                 )
+    ):
+        root = Path(os.environ.get("VIRTUAL_ENV"), "..").resolve()
+    else:
+        root = Path("~/invokeai").expanduser().resolve()
+    return root
+
+class InvokeAIAppConfig(InvokeAISettings):
+    '''
+    Application-wide settings.
+    '''
+    #fmt: off
+    type: Literal["globals"] = "globals"
+    root                : Path = Field(default=_find_root(), description='InvokeAI runtime root directory')
+    infile              : Path = Field(default=None, description='Path to a file of prompt commands to bulk generate from')
```

Comment:
Is this still functional?

Reply:
The `root` is functional. `infile` is not, but preserved for future compatibility.

Reply:
Maybe comment out infile if it's not functional?

Reply:
Sure. Will do.